dbms/src/Storages/S3/FileCache.cpp

```diff
@@ -166,6 +233,69 @@ FileSegmentPtr FileCache::get(const S3::S3FilenameView & s3_fname, const std::op
     return nullptr;
 }
 
+FileSegmentPtr FileCache::getOrWait(const S3::S3FilenameView & s3_fname, const std::optional<UInt64> & filesize)
+{
+    auto s3_key = s3_fname.toFullKey();
+    auto file_type = getFileType(s3_key);
+    auto & table = tables[static_cast<UInt64>(file_type)];
+
+    std::unique_lock lock(mtx);
+
+    auto f = table.get(s3_key);
+    if (f != nullptr)
+    {
+        lock.unlock();
+        f->setLastAccessTime(std::chrono::system_clock::now());
+        auto status = f->waitForNotEmpty();
+        if (status == FileSegment::Status::Complete)
+        {
+            GET_METRIC(tiflash_storage_remote_cache, type_dtfile_hit).Increment();
+            return f;
+        }
+        else
+        {
+            // On-going download failed, let the caller retry.
+            return nullptr;
+        }
+    }
+
+    GET_METRIC(tiflash_storage_remote_cache, type_dtfile_miss).Increment();
+
+    auto estimated_size = filesize ? *filesize : getEstimatedSizeOfFileType(file_type);
+    if (!reserveSpaceImpl(file_type, estimated_size, /*try_evict*/ true))
+    {
+        // Space not enough.
+        GET_METRIC(tiflash_storage_remote_cache, type_dtfile_full).Increment();
+        LOG_DEBUG(
```

Comment:
Why this log level is DEBUG here? It throws later, why not append this log into exception itself, or change to INFO level?

Reply:
The exception will be caught by the caller, will not print to log.

Reply:
If the caller hits `i <= 1`, then it will be thrown?

Reply:
so the log is necessary.

Reply:
If the exception is thrown, and the log level is INFO which is normal, then we can get no information like `(capacity={} used={} estimzted_size={})`

Reply:
fixed